### DISCUSSION TITLE
Question: any helm chart available?

### DISCUSSION THREAD

**Discussion by junquero** [arkmq-org/arkmq-org-broker-operator#788] [Q&A]
Is there any helm chart available to install the artemis operator instead of using the install scripts?

If not, is there any plan to generate that helm charts and publish it in https://artifacthub.io/ or any other public helm repository?

**Comment by brusdev** [comment]
There are no plans to generate helm charts but the OLM bundle is already available and there are already plans to publish the ArtemisCloud Operator in [operatorhub.io](https://operatorhub.io/) as an OLM operator. Feel free to create a new issue to publish the ArtemisCloud Operator in https://artifacthub.io/ too.

**Reply by mycroft11** [reply]
That's good news, any timelines for this?

**Reply by brusdev** [reply]
@mycroft11 I should be able to provide updates on timelines the next week.

**Reply by mycroft11** [reply]
Hi @brusdev , looking forward to seeing this on operatorhub, any updates?

**Comment by gw-kamara** [comment]
@junquero - Literally just done this today
copy the yaml files listed in the deploy.sh file into your helm templates directory and you're done. 

address_queue.yaml - (from the examples/address/address_queue.yaml)
artemis_single.yaml - (from examples/artemis/artemis_single.yaml) 
cluster_role.yaml
cluster_role_binding.yaml
crds - (whole directory) 
election_role.yaml
election_role_binding.yaml
operator.yaml
role.yaml
role_binding.yaml
service_account.yaml

**Reply by junquero** [reply]
@gw-kamara, yes, I did it too, but there are a couple of issues you have to take into account:
1. crds directory can't be under the templates directory. It has to be at the same level than the templates directory.
2. you can have two helm charts: one for cluster-wide and another for namespaced.

I have done it and I have my own helm charts running. I asked the question because if they are interested in supporting helm charts, I can provide what I did... but it looks like they are not interested in supporting it, because they prefer to support the OLM bundle.

**Reply by brusdev** [reply]
@junquero I didn't mean to discourage you. If supporting the helm charts doesn't require a big maintenance effort, feel free to create a pull request and to set me as reviewer. Any contributions is always welcome.

**Reply by junquero** [reply]
@brusdev thank you for your new approach. The point here is not creating the helm chart(s) which is a quite simple task, the important thing is to decide which helm repo is going to be published the charts. Also, to integrate the helm chart build in the build and deploy pipeline.

I will create the pull request anyway, because it is a quite simple shell script, and then you can decide what to do with it.

**Reply by teksuo** [reply]
Hi there, any news on this?

**Reply by brusdev** [reply]
@teksuo there are no plans to generate helm charts because the OLM bundle is already available and there are already plans to publish the ArtemisCloud Operator in [operatorhub.io](https://operatorhub.io/) as an OLM operator but anycontributions to support helm charts is welcome.

**Comment by Stevenpc3** [comment]
We have been using helm-charts for deployment for many years.  It is largely baked into out automation.  The current method of deploying this operator tool is NOT very friendly or automated.  Adopting OLM is also NOT something we will be able to do any time soon as it will require review and approval.

We deliver our overall tools  to a customer in special environments and it requires us to have items packaged and automated as much as possible while providing config override files for different environments.  We have to be "hands off" as much as possible and almost all aspects are automated.  Artemis is only 1 of the 100+ services (custom and FOSS) we package as part of our very large integrated application.

We have created our own very robust Artemis helm-chart and make our own docker image that (I would love to share) has served us well over the years.  We have been waiting for some "cloud" version of Artemis for a very long time and we have had to solve many issues that this repo is just now getting to and some it appears only we have solved...  We would love to contribute and adopt this solution, but we cannot use it as is without extensive work.

Creating a helm-chart would allow us to start testing with this and then we can funnel back fixes and ideas and issues.  There are several good options to compare a proper chart with that use operators too.

Here are a few good operators (all we use) that have helm-charts as part of their deployment offerings.

1. https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack
2. https://github.com/elastic/cloud-on-k8s/tree/main/deploy
3. https://github.com/cloudnative-pg/charts
4. https://github.com/jaegertracing/helm-charts/tree/main/charts/jaeger-operator

The general practice is to install the operator and have another chart to install the "config" that the operator uses.  I get that we can make a PR, but the integration into the operator code is critical for a good solution.  Just moving *.yaml files to a helm chart then doing your own thing is not great.

For the charts listed above we create our own "wrapper" or "umbrella" charts where we list thier charts as our dep.  This allows us to build templates and customize our values.yamls as overrides eaiser.  We likely will have to use kustomize for some like CNPG if they do not expand their options or take more PRs.  ECK and kube-prom-stack are **exceptional** examples of highly customizable helm charts that offer operators and instances. 

There are certainly CRD concerns to be had and that is covered more in the kube docs and various examples in other charts and this is not a super light load, but it would be greatly beneficial in encouraging others to adopt the tool.

Here is another reference about templating usage and how it's beneficial as part of helm-charts (could also combine with Kustomize) https://github.com/cloudnative-pg/charts/pull/346#issuecomment-2277986527

**Reply by brusdev** [reply]
@Stevenpc3 thanks for your feedback, I see the benefits of using helm-charts and I'm not against them but there are no volunteers working on it at the moment and I don't know how they could fit in the ArtemisCloud Operator.
Could you share your Artemis helm-chart to deepen your expectations?

**Reply by Stevenpc3** [reply]
I will be working on sharing the image and helm chart we have been using for years and adding to, but the issue is that it is not for operators.

In the above comment I linked several projects that offer helm charts for operators.  Generally there are two charts.  One is for deploying the operator and crds and the second is for deploying a config/resource that is used to tell the operator what to do.  For example the ECK operator https://github.com/elastic/cloud-on-k8s/tree/main/deploy you would deploy the operator first, then follow up with the eck-stack.

So while the charts we use will help with expectations on customization, it will not help with the format of a good set of charts for operators.

We would love to move to this but again, we cannot adopt the OLM at this time.  Having helm charts would go a long way for the community.

**Reply by Stevenpc3** [reply]
@brusdev I made an example of our charts here https://github.com/Stevenpc3/artemis-helm-charts/tree/main it outlines the layout and usage of how we deploy artemis to Kubernetes now and for the last 5 years?  We actually had to invent quite a few things on our own and then wait for Artemis devs to make updates we could use.

I hope it is helpful.  If there are any questions please reach out.  I would love to help how I can.

Is there a way for me to make issues or requests to Artemis itself?

**Reply by brusdev** [reply]
@Stevenpc3 thanks for sharing an example of your charts. Your approach is interesting but I'm not sure it could fit in the ArtemisCloud Operator, it seems an alternative approach to the [operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/). While the approach of the [ECK operator](https://github.com/elastic/cloud-on-k8s/tree/main/deploy) uses helm charts to install an operator. 

> Is there a way for me to make issues or requests to Artemis itself?
Sure, feel free to create new issues at https://github.com/arkmq-org/activemq-artemis-operator/issues

Of course any contributions is welcome 🙂

**Reply by Stevenpc3** [reply]
Yes, I stated above that our chart is NOT for operators. 

I also showed many other good operator chat examples. 

The chart I shared only shows what we currently expect to be able to customize which is good insight for even the operator. 

If we can't set those things using the operator (even if there was a helm chat or not) then we would not be able to use the operator.

For contribution, I mean the Artemis code request. This is I guess what I need https://activemq.apache.org/issues

**Reply by brusdev** [reply]
@Stevenpc3 I see your point. It makes sense to me. I did a review of the settings defined at https://github.com/Stevenpc3/artemis-helm-charts/blob/main/charts/artemis/README.md#values. The operator already supports the customization of all those settings except the oAuthSidecar and the zookeeper-bitnami features by using the fields in the [ActiveMQArtemis CRD](https://doc.crds.dev/github.com/arkmq-org/activemq-artemis-operator/broker.amq.io/ActiveMQArtemis/v1beta1@v2.0.0), the [broker-properties](https://github.com/arkmq-org/activemq-artemis-operator/blob/main/docs/help/operator.md#configuring-brokerproperties) and the [resource templates](https://github.com/arkmq-org/activemq-artemis-operator/blob/main/docs/help/operator.md#custom-modifications-via-a-strategic-merge-patch).

The oAuthSidecar is an interesting feature but we plan to externalize the web console by using [hawtio-online](https://github.com/hawtio/hawtio-online).

How does the zookeeper-bitnami feature work? Is a zookeeper cluster shared across multiple broker clusters?

> For contribution, I mean the Artemis code request. This is I guess what I need https://activemq.apache.org/issues
Indeed, it is.

**Reply by Stevenpc3** [reply]
Good to hear about the configuration. Thank you for the elaboration.

I will read more about hawtio-online and compare to oauth2-proxy. We just need to authenticate the management console and it integrates with keycloak and many others .

For zookeeper we follow https://activemq.apache.org/components/artemis/documentation/latest/ha.html and posts like https://stackoverflow.com/questions/78495784/integration-activemq-artemis-with-zookeeper-does-not-work-on-colocation with posts from Justin Bertram THE Artemis maintainer/contributor to handle "split-brain" in our 2 instance cluster. It's extremely dangerous and we need HA but cannot deploy in a 3*3 quorum due to resources. So this provides that protection. 

I hope the operator provides HA and handles split -brain.

**Reply by brusdev** [reply]
> I will read more about hawtio-online and compare to oauth2-proxy. We just need to authenticate the management console and it integrates with keycloak and many others .

The new Apache ActiveMQ Artemis console is based on hwatio 4 that supports generic OpenID authentication, see https://hawt.io/docs/oidc.html.

> I hope the operator provides HA and handles split -brain.

ArkMQ Operator supports shared store HA that doesn't require 3*3 quorum to handle split -brain because it uses DB or FS locks, see the following tests:
https://github.com/arkmq-org/activemq-artemis-operator/blob/main/controllers/activemqartemis_jdbc_ha_test.go
https://github.com/arkmq-org/activemq-artemis-operator/blob/main/controllers/activemqartemis_rwm_pvc_ha_test.go

**Comment by jasaltvik** [comment]
@brusdev Hi, any update on your plans to deploy to operatorhub.io? Would be really cool to have a community version of this available in OpenShift 💯We are potentially planning on migrating to this one from the Red Hat one, and it would be really cool if we could just switch via operatorhub. Thanks for any updates on this

**Reply by Msarawan** [reply]
Hello @jasaltvik , I’m currently working on it, following the steps outlined in the [operatorhub.io](https://operatorhub.io/) documentation for publishing our operator. At the moment, I’m testing the operator using OLM to ensure it meets the required criteria. Once the testing is complete, I’ll proceed with creating a pull request on the Community Operators GitHub repository. I’ll continue to share updates as progress is made. Thank you!

**Reply by jasaltvik** [reply]
Thanks, great news! Looking forward to see it on operatorhub!

**Reply by jasaltvik** [reply]
@Msarawan Hey, just checking in to see if there's any progress on this? Anywhere I can follow progress on this? Thanks for any updates.

**Reply by Msarawan** [reply]
Hello @jasaltvik , I initially began testing our operator with OLM but paused that work after encountering issues related to the old SQLite-based catalog. To address this, I shifted focus to migrating from the deprecated SQLite-based catalog to the newer File-Based Catalog (FBC) format. This involved updating our automation and CI setup, which took some time.

Now i have a PR ready that includes all these changes. Once it’s merged and we complete the upstream operator release, I’ll move forward with submitting the PR to the Community Operators repository.

I expect to need another 2–3 weeks to finalize everything. I’ll keep you posted as progress continues — thanks again for your patience!

**Reply by jasaltvik** [reply]
@Msarawan No worries, and thanks for the informative update! And thanks for working on this

**Reply by jasaltvik** [reply]
@Msarawan Hi again, just checking what the progress is on this? I probably need to migrate during next week, and it would be awesome if I could just move over to a community operator instead of having to figure out something else.

**Reply by jasaltvik** [reply]
> @Msarawan Hi again, just checking what the progress is on this? I probably need to migrate during next week, and it would be awesome if I could just move over to a community operator instead of having to figure out something else.

I assume it needs to be deployed here: https://github.com/redhat-openshift-ecosystem/community-operators-prod ?

**Reply by Msarawan** [reply]
Hello @jasaltvik , Yes, I made progress, and I’m happy to share that our upstream ArkMQ Operator is now published on [operatorhub.io](https://operatorhub.io/operator/activemq-artemis-operator)!  
Now, users can install the ArkMQ Operator in a Kubernetes cluster with OLM by creating a simple subscription.

Regarding the community OpenShift Operators, yes — we plan to publish it on [community-operators-prod](https://github.com/redhat-openshift-ecosystem/community-operators-prod) as well. I’m just waiting on final confirmation from the team, and once that’s in, our operator should be available there by the end of next week.

I’ll keep you updated as soon as it’s published.

**Reply by Msarawan** [reply]
Hello @jasaltvik , 
I’m happy to share that our upstream operator has now been published and is officially available in the OpenShift OperatorHub. 
You can now install it directly from the OpenShift console to test the latest upstream features and bug fixes.
<img width="1917" height="1013" alt="Screenshot From 2025-09-02 21-49-34" src="https://github.com/user-attachments/assets/e39cccd1-dad8-4948-b78c-fcf7cf25022e" />

**Reply by jasaltvik** [reply]
> Hello @jasaltvik , 
> I’m happy to share that our upstream operator has now been published and is officially available in the OpenShift OperatorHub. 
> You can now install it directly from the OpenShift console to test the latest upstream features and bug fixes.
> <img width="1917" height="1013" alt="Screenshot From 2025-09-02 21-49-34" src="https://github.com/user-attachments/assets/e39cccd1-dad8-4948-b78c-fcf7cf25022e" />
> 

Awesome! Thank you so much!

**Reply by Msarawan** [reply]
Hello @jasaltvik ,
Just to keep you informed — we have removed the old entry of our operator from the Kubernetes Community [Operatorhub](https://operatorhub.io/) as we planned to update the package name with our upstream `arkmq-org` prefix. We are now re-submitting the operator under the new package name.
Here is relevant issue for this change: https://github.com/arkmq-org/activemq-artemis-operator/issues/1209

**Reply by jasaltvik** [reply]
> Hello @jasaltvik , Just to keep you informed — we have removed the old entry of our operator from the Kubernetes Community [Operatorhub](https://operatorhub.io/) as we planned to update the package name with our upstream `arkmq-org` prefix. We are now re-submitting the operator under the new package name. Here is relevant issue for this change: #1209

@Msarawan Thanks for the heads-up. We are using it through OpenShift in production. I just saw https://github.com/redhat-openshift-ecosystem/community-operators-prod/pull/7650. How will this affect running instances? Thanks for any input

**Reply by brusdev** [reply]
@jasaltvik the running instances will not be affected. It will affect the upgrade procedure. When the new entry with the `arkmq-org prefix` is published, you need to manually uninstall the operator (without uninstalling the CRDs) and install the operator with the `arkmq-org` prefix. WIth this procedure, the running broker pods will not be affected.

**Reply by jasaltvik** [reply]
@brusdev Thanks. Do you have an ETA on when you might be merging https://github.com/redhat-openshift-ecosystem/community-operators-prod/pull/7650, and re-submitting it with the new name?

**Reply by brusdev** [reply]
> @brusdev Thanks. Do you have an ETA on when you might be merging [redhat-openshift-ecosystem/community-operators-prod#7650](https://github.com/redhat-openshift-ecosystem/community-operators-prod/pull/7650), and re-submitting it with the new name?

We have no ETA yet. We will update this discussion ASAP.

**Reply by jasaltvik** [reply]
@brusdev / @Msarawan So it seems you have deleted the operator now as I can see errors on the currently running `Subscription` in all environments. It looks something like this

```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: activemq-artemis-operator
  namespace: my-namespace
spec:
  channel: stable
  installPlanApproval: Automatic
  name: activemq-artemis-operator
  source: community-operators
  sourceNamespace: openshift-marketplace
```

Did you deploy the replacement yet? I cannot seem to find anything.

**Reply by Msarawan** [reply]
Hello @jasaltvik, Yes, I created the PRs today for republishing the operator. The PR on the Community Kubernetes Operators repository has been merged, and the operator is now available on [OperatorHub.](https://operatorhub.io/operator/arkmq-org-broker-operator)
I’ve also created the same PR in the openshift-community-operators repository [PR#7863](https://github.com/redhat-openshift-ecosystem/community-operators-prod/pull/7863) All the CI checks have passed; I’m just waiting for the PR to be merged. Once it’s merged, I’ll let you know.
Thank you!
